### PR TITLE
Improve readability for `interfaces/jax_jit_tuple.py`

### PR DESCRIPTION
**Pull request by albi3ro**
For ADR PR #67 https://github.com/PennyLaneAI/adrs/pull/67 with prototype PR #4368 I have been spending time going through `jax_jit_tuple.py` to figure out if and how we can integrate device jacobian products.

To better understand the file, I have made some stylistic improvements. So though a lot of things are moved around, the code is the exact same.

Some of these stylistic improvements include:

* A few more docstrings with examples and adding some explicit documentation on closure variables to docstrings.

* Eliminating some single line functions: `process_single_shape` and `_filter_zero_tangents`

* Switching some iterates and append patterns with function + tuple with generator pattern.

* Switch `_jac_shape_dtype_tuple` for a batch of tapes with `jac_shape_dtype_struct` for a single tape. The code that uses can then determine how it wants to nest a batch with a single tape.

* Extract out `_grad_transform_jac_via_callaback` and `_device_method_jac_via_callback` to being global functions with explicit argument lists, instead of depending on closure. This reduces nesting, allows the functions to be experimented on and tested independently, and makes the dependencies explicit. The previous syntax also called local variables like `params` the same thing as a parameter from outer scope.

* Extract a logical branch into `_grad_transform_jac_no_callback`. Then the three options all have the same syntax pattern.

* Extract code from the forward mode `execute_wrapper_jvp` into `_empty_jacs_for_shape` and `_switched_jacobian`.  I will freely admit I'm confused by the code of `_switched_jacobian`, but it works and passes tests.

* Got rid of the `switch_trainable` list to instead just check that quantity in each loop, instead of precomputing it. This way we don't have to remember what `switch_trainable` represents.

* A few more minor things I've forgotten about.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -65,10 +65,12 @@
 * The experimental device interface is integrated with the `QNode` for Jax jit.
   [(#4352)](https://github.com/PennyLaneAI/pennylane/pull/4352)
 
-* Added functions `adjoint_jvp` and `adjoint_vjp` to `qml.devices.qubit.preprocess` that computes
+* Added functions `adjoint_jvp` and `adjoint_vjp` to `qml.devices.qubit.adjoint_jacobian` that computes
   the JVP and VJP of a tape using the adjoint method.
   [(#4358)](https://github.com/PennyLaneAI/pennylane/pull/4358)
 
+* Readability improvements and stylistic changes to `pennylane/interfaces/jax_jit_tuple.py`
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:72]:
```suggestion
* Readability improvements and stylistic changes to `pennylane/interfaces/jax_jit_tuple.py`
  [(#4379)](https://github.com/PennyLaneAI/pennylane/pull/4379/)
```

File: pennylane/interfaces/jax_jit_tuple.py [hunk 2]
```diff
@@ -43,71 +44,187 @@ def set_parameters_on_copy_and_unwrap(tapes, params, unwrap=True):
...
-    shape_dtypes = []
+    shape_and_dtype = _create_shape_dtype_struct(tape, device)
+    if len(tape.trainable_params) == 1:
+        return shape_and_dtype
+    if len(tape.measurements) == 1:
+        return tuple(shape_and_dtype for _ in tape.trainable_params)
+    return tuple(tuple(_s for _ in tape.trainable_params) for _s in shape_and_dtype)
+
+
+def _empty_jacs_for_shape(
```
Comment (albi3ro) [inline pennylane/interfaces/jax_jit_tuple.py:84]: See master line 422

File: pennylane/interfaces/jax_jit_tuple.py [hunk 3]
```diff
@@ -43,71 +44,187 @@ def set_parameters_on_copy_and_unwrap(tapes, params, unwrap=True):
...
+    if isinstance(shape_dtype, tuple):
+        return tuple(_empty_jacs_for_shape(struct) for struct in shape_dtype)
+    return jnp.zeros(shape_dtype.shape, shape_dtype.dtype)
+
 
-    for t in tapes:
-        shape_and_dtype = _create_shape_dtype_struct(t, device)
-        shape_dtypes.append(shape_and_dtype)
-    return tuple(shape_dtypes)
+def _switched_jacobian(tape, device, original_trainable_parameters, jacs):
```
Comment (albi3ro) [inline pennylane/interfaces/jax_jit_tuple.py:96]: See master lines 418-423

File: pennylane/interfaces/jax_jit_tuple.py [hunk 4]
```diff
@@ -43,71 +44,187 @@ def set_parameters_on_copy_and_unwrap(tapes, params, unwrap=True):
...
         tape_dtype = _numeric_type_to_dtype(tape.numeric_type)
-        return process_single_shape(shape, tape_dtype)
+        return jax.ShapeDtypeStruct(tuple(shape), tape_dtype)
 
     tape_dtype = tuple(_numeric_type_to_dtype(elem) for elem in tape.numeric_type)
-    return tuple(process_single_shape(s, d) for s, d in zip(shape, tape_dtype))
+    return tuple(jax.ShapeDtypeStruct(tuple(s), d) for s, d in zip(shape, tape_dtype))
 
 
-def _tapes_shape_dtype_tuple(tapes, device):
```
Comment (albi3ro) [inline pennylane/interfaces/jax_jit_tuple.py:63]: This function can be reduced to a single line with a tuple + generator.  So it is eliminated in favor of 
```
tuple(_create_shape_dtype_struct(t, device) for t in tapes)
```

**Review by timmysilv — APPROVED** [review]
much much nicer. left one comment for curiosity more than anything, but otherwise lgtm!

File: pennylane/interfaces/jax_jit_tuple.py [hunk 5]
```diff
@@ -43,71 +44,187 @@ def set_parameters_on_copy_and_unwrap(tapes, params, unwrap=True):
...
+
+def _empty_jacs_for_shape(
+    shape_dtype: Union[Tuple, jax.ShapeDtypeStruct]
+) -> Union[Tuple, jnp.array]:
+    """Converts a nested tuple containing ``jax.ShapeDtypeStruct`` into a nested tuples
+    containing zeros arrays of the appropriate shape.
+
+    """
+    if isinstance(shape_dtype, tuple):
+        return tuple(_empty_jacs_for_shape(struct) for struct in shape_dtype)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jax_jit_tuple.py:92]: Both recursion and improving the level of abstraction.

For the original code:
```
# Adapt the shape of the empty jacobian given the measurement shape
shape_dtype = original_shape if len(tapes) == 1 else original_shape[i]
if isinstance(shape_dtype, tuple):
    shape_dtype = shape_dtype[0]

# Multi measurement
if multi_measurement:
    jac_empty = tuple(
        jnp.zeros(shape=tensor.shape, dtype=tensor.dtype) for tensor in shape_dtype
    )
# Single measurement
else:
    jac_empty = jnp.zeros(shape_dtype.shape, shape_dtype.dtype)
```

I was struggling with "how nested is the tuple at this point in time".  In what circumstances was `shape_dtype` a tuple that we needed to eliminate a nested dimension from? What dimension were we eliminating in that case?

I always ended up either having `shape_dtype` not be a tuple under `multi_measurement`, or having it be a tuple under single measurements.

But this recursion pattern seemed to work.

**Pull request merged**